actors/builtin/market/market_actor.go

```diff
@@ -181,94 +180,160 @@ func (a Actor) PublishStorageDeals(rt Runtime, params *PublishStorageDealsParams
 	}
-
 	resolvedAddrs := make(map[addr.Address]addr.Address, len(params.Deals))
 	baselinePower := requestCurrentBaselinePower(rt)
 	networkRawPower, networkQAPower := requestCurrentNetworkPower(rt)
 
-	var newDealIds []abi.DealID
+	// Drop invalid deals
 	var st State
+	proposalCidLookup := make(map[cid.Cid]struct{})
+	validProposalCids := make([]cid.Cid, 0)
+	validDeals := make([]ClientDealProposal, 0, len(params.Deals))
+	totalClientLockup := make(map[addr.Address]abi.TokenAmount)
+	totalProviderLockup := abi.NewTokenAmount(0)
+
+	validInputBf := bitfield.New()
+	rt.StateReadonly(&st)
+	msm, err := st.mutator(adt.AsStore(rt)).withPendingProposals(ReadOnlyPermission).
+		withEscrowTable(ReadOnlyPermission).withLockedTable(ReadOnlyPermission).build()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load state")
+	for di, deal := range params.Deals {
+		/*
+			drop malformed deals
+		*/
+		if err := validateDeal(rt, deal, networkRawPower, networkQAPower, baselinePower); err != nil {
+			rt.Log(rtt.INFO, "invalid deal %d: %s", di, err)
+			continue
+		}
+		if deal.Proposal.Provider != provider && deal.Proposal.Provider != providerRaw {
+			rt.Log(rtt.INFO, "invalid deal %d: cannot publish deals from multiple providers in one batch", di)
+			continue
+		}
+		client, ok := rt.ResolveAddress(deal.Proposal.Client)
+		if !ok {
+			rt.Log(rtt.INFO, "invalid deal %d: failed to resolve proposal.Client address %v for deal ", di, deal.Proposal.Client)
+			continue
+		}
+
+		/*
+			drop deals with insufficient lock up to cover costs
+		*/
+		if _, ok := totalClientLockup[client]; !ok {
+			totalClientLockup[client] = abi.NewTokenAmount(0)
+		}
+		totalClientLockup[client] = big.Sum(totalClientLockup[client], deal.Proposal.ClientBalanceRequirement())
+		clientBalanceOk, err := msm.balanceCovered(client, totalClientLockup[client])
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check client balance coverage")
+		if !clientBalanceOk {
+			rt.Log(rtt.INFO, "invalid deal: %d: insufficient client funds to cover proposal cost", di)
+			continue
+		}
+		totalProviderLockup = big.Sum(totalProviderLockup, deal.Proposal.ProviderCollateral)
+		providerBalanceOk, err := msm.balanceCovered(provider, totalProviderLockup)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check provider balance coverage")
+		if !providerBalanceOk {
+			rt.Log(rtt.INFO, "invalid deal: %d: insufficient provider funds to cover proposal cost", di)
+			continue
+		}
+
+		/*
+			drop duplicate deals
+		*/
+		// Normalise provider and client addresses in the proposal stored on chain.
+		// Must happen after signature verification and before taking cid.
+		deal.Proposal.Provider = provider
+		resolvedAddrs[deal.Proposal.Client] = client
+		deal.Proposal.Client = client
+
+		pcid, err := deal.Proposal.Cid()
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalArgument, "failed to take cid of proposal %d", di)
```

Comment:
I'm not sure that we'd want to, but we could just log and continue from errors such as this (and the many other unexpected errors).

Reply:
This is a good thing to raise.  As far as I can tell this and the many illegal state errors represent programming errors and I want to keep those aborting the whole thing.

Reply:
makes sense, i think i agree